design/02_On_Chain_Quotes.md

```diff
@@ -0,0 +1,199 @@
+    ExecutorQuoterRouter-->>Integrator: value
+    Integrator-->>OffChain: value
+    OffChain->>Integrator: execute{value}(...)
+    Integrator->>ExecutorQuoterRouter: requestExecution{value}(quoterAddr, ...)
+    ExecutorQuoterRouter->>ExecutorQuoterRouter: lookupQuoterImplementation
+    ExecutorQuoterRouter->>ExecutorQuoter: quote(...)
+    ExecutorQuoter->>ExecutorQuoter: custom logic
+    ExecutorQuoter-->>ExecutorQuoterRouter: payee, value
+    ExecutorQuoterRouter->>ExecutorQuoterRouter: buildQuote
+    ExecutorQuoterRouter->>Executor: requestExecution{value}
+    ExecutorQuoterRouter->>ExecutorQuoterRouter: emit OnChainQuote
+```
+
+# Detailed Design
+
+The existing Executor contracts are immutable, handle payment in the native gas token, and require a standardized quote header. This design introduces and standardizes the minimum viable approach to form quotes on-chain, allow for permissionless quoter selection, and reuse the rest of the on- and off-chain tooling.
+
+## Technical Details
+
+### EVM
+
+On EVM, two new contracts will be introduced.
+
+1. **ExecutorQuoter** represents the on-chain quoting logic of a particular Quoter / Relay Provider. It may implement any logic desired by the Relay Provider as long as it adheres to this interface. It SHOULD be immutable.
+
+   ```solidity
+   interface IExecutorQuoter {
+       function requestQuote(
+           uint16 dstChain,
+           bytes32 dstAddr,
+           address refundAddr,
+           bytes calldata requestBytes,
+           bytes calldata relayInstructions
+       ) external view returns (bytes32 payeeAddress, uint256 requiredPayment);
+   }
+   ```
+
+2. **ExecutorQuoterRouter** replaces **Executor** as the entry-point for integrators. It MUST be immutable and non-administered / fully permissionless. This provides three critical functionalities.
+
+   1. `updateQuoterContract(bytes calldata gov)` allows a Quoter to set their `ExecutorQuoter` contract via signed governance (detailed below). This MUST
+      1. Verify the chain ID matches the Executor’s `ourChain`.
+      2. Verify the contract address is an EVM address.
+      3. Verify the sender matches the sender on the governance.
+      4. Verify the governance has not expired.
+      5. Verify the signature `ecrecover`s to the quoter address on the governance.
+      6. Assign the specified contract address to that quoter address.
+      7. Emit a `QuoterContractUpdate` event.
+   2. `quoteExecution` allows an integrator to quote the cost of an execution for a given quoter in place of a signed quote. This MUST call `requestQuote` from that Quoter’s registered contract.
+   3. `requestExecution` allows an integrator to request execution via Executor providing a quoter address in place of a signed quote. This MUST
+      1. Call `requestQuote` from that Quoter’s registered contract.
+      2. Enforce the required payment.
+      3. Refund excess payment.
+      4. Request execution, forming a `EQ02` quote on-chain.
```

Comment:
Should we show what an `EQ02` quote looks like here?

Reply:
```suggestion
      4. Request execution, forming a [`EQ02`](#quote---version-2) quote on-chain.
```

Reply:
I'm good with linking since it's in the below section where those things typically go